---
title: "Pathway Workflow"
output: 
  html_document:
    code_folding: hide
params:
  pre_processed:
  sample_info:
  all_gene_info:
  deg:
  idep_data:
  converted:
  all_gene_names:
  go:
  select_org:
  my_range:
  select_contrast:
  min_set_size:
  max_set_size:
  limma:
  gene_p_val_cutoff:
  gene_sets:
  absolute_fold:
  pathway_p_val_cutoff:
  n_pathway_show:
  contrast_samples:
  sig_pathways:
  #heatmap_color_select:
  pathway_method:
  #gage_pathway_data:
  #selected_pathway_data:
  pathway_list_data:
  #network_data:
  #pgsea_plot_data:
  # fgsea_pathway_data:
  # pgsea_plot_data:
  # pgsea_plot_all_samples_data:
  # gene_info:
  date:
    label: "Date: "
    value: !r Sys.Date()
  printcode:
    label: "Display Code"
    value: TRUE
    input: checkbox
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = params$printcode) # , include = FALSE, eval = FALSE)
```

<!-- This document cannot be knit directly from R studio -->
<div>

::: {style="color: Blue"}
## If this site has contributed to your work, please cite our article:

Ge, S.X., Son, E.W. & Yao, R. iDEP: an integrated web application for differential expression and pathway analysis of RNA-Seq data. BMC Bioinformatics 19, 534 (2018). <https://doi.org/10.1186/s12859-018-2486-6>
:::

## 

</div>
This html document contains what parameters values were selected on the IDEP interface. It also includes the plots generated from those selections. Genes with FDR Greater than **`r params$gene_p_val_cutoff`** were removed.


# Summary

The comparison selected was **`r params$select_contrast`** using **`r params$go`** genesets. **`r params$n_pathway_show`** pathways with between **`r params$min_set_size`** and **`r params$max_set_size`** genes were included. Absolute fold values were `r ifelse(params$absolute_fold, 'not', '')`  used. Pathways were found using the **`r switch(as.character(params$pathway_method), "1" = "GAGE", "2" = "PGSEA", "3" = "GSEA (Pre-Ranked)", "4" = "PGSEA with all samples", "5" = "ReactomePA" )`** method. Pathway p values cutoff is set to **`r params$pathway_p_val_cutoff`**.



<!--ALL SELECTED PARAMETERS-->
# Selections
```{r echo=FALSE}

# exclude long parameter we won't print
excluded_params <- c(
  "sample_info", "loaded_data", "pre_processed", "deg",
  "idep_data", "converted", "limma", "gene_sets",
  "pathway_list_data", "all_gene_names", "all_gene_info"
)

for (i in 1:length(params)) {
  if (!(names(params)[i] %in% excluded_params)) {
    cat(paste0(names(params)[i], ": ", params[[i]], "\n"))
  }
}
```

# Tree
```{r echo=FALSE}
enrichment_tree_plot(
  go_table = params$pathway_list_data,
  group = "All Groups",
  right_margin = 10
)
```

```{r eval=FALSE, include=FALSE}

# GAGE Method
if (params$pathway_method == 1) {
  gene_sets <- read_gene_sets(
    converted = params$converted,
    all_gene_names = params$all_gene_names,
    go = params$go,
    select_org = params$select_org,
    idep_data = params$idep_data,
    my_range = params$my_range
  )

  gage <- gage_data(
    select_go = params$go,
    select_contrast = params$select_contrast,
    min_set_size = params$min_set_size,
    max_set_size = params$max_set_size,
    limma = params$limma,
    gene_p_val_cutoff = params$gene_p_val_cutoff,
    gene_sets = params$gene_sets,
    absolute_fold = params$absolute_fold,
    pathway_p_val_cutoff = params$pathway_p_val_cutoff,
    n_pathway_show = params$n_pathway_show
  )

  # basic_heatmap(
  #   data = params$selected_pathway_data,
  #   c("green", "black", "red")
  # )

  # vis_network_plot(
  #     network_data = params$network_data
  # )


  # enrichment_tree_plot(
  #   go_table = params$pathway_list_data,
  #   group = "All Groups"
  #   #45
  # )
}
```

```{r eval=FALSE, include=FALSE}
# PGSEA -- Not tested
if (params$pathway_method == 2) {
  res <- get_pgsea_plot_data(
    my_range = c(params$min_set_size, params$max_set_size),
    data = params$pre_process$data,
    select_contrast = params$select_contrast,
    gene_sets = params$gene_sets,
    sample_info = params$sample_info,
    select_factors_model = deg$select_factors_model,
    select_model_comprions = deg$select_model_comprions,
    pathway_p_val_cutoff = params$pathway_p_val_cutoff,
    n_pathway_show = params$n_pathway_show
  )

  pathway_list_data <- get_pathway_list_data(
    pathway_method = params$pathway_method,
    gage_pathway_data = NULL, # gage_pathway_data(),
    fgsea_pathway_data = NULL, # fgsea_pathway_data(),
    pgsea_plot_data = res,
    pgsea_plot_all_samples_data = NULL, # pgsea_plot_all_samples_data(),
    go = params$go,
    select_org = params$select_org,
    gene_info = params$all_gene_info,
    gene_sets = params$gene_sets
  )

  enrichment_tree_plot(
    go_table = pathway_list_data,
    group = "All Groups" # ,
    # right_margin = 45
  )
}
```
