---
title: "Principal Component Analysis Workflow"
output: 
  html_document:
    code_folding: hide
params:
  pre_processed_data:
  pre_processed_descr:
  sample_info:
  pc_x:
  pc_y:
  pc_z:
  color:
  shape:
  all_gene_names:
  select_gene_id:
  selected_x:
  selected_y:
  encircle:
  showLoadings:
  pointlabs:
  point_size:
  ui_color:
  ui_shape:
  plots_color_select:
  printcode:
    label: "Display Code"
    value: TRUE
    input: checkbox
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = params$printcode, warning = FALSE, message = FALSE)
```


```{r include=FALSE}
# load package
library(idepGolem)
```

<!-- This document cannot be knit directly from R studio -->
<div>

::: {style="color: Blue"}
## If this site has contributed to your work, please cite our article:

Ge, S.X., Son, E.W. & Yao, R. iDEP: an integrated web application for differential expression and pathway analysis of RNA-Seq data. BMC Bioinformatics 19, 534 (2018). <https://doi.org/10.1186/s12859-018-2486-6>
:::

## 

</div>

This html document contains what parameters values were selected on the IDEP interface of the "PCA" tab. It also includes the plots generated from those selections.

## Pre-Process Summary
`r params$pre_processed_descr`

## All PCA inputs
```{r echo=FALSE}
for (i in 1:length(params)) {
  # exclude loaded data & sample info
  if (names(params)[i] != "sample_info" && names(params)[i] != "pre_processed_data" && names(params)[i] != "all_gene_names") {
    cat(paste0(names(params)[i], ": ", params[[i]], "\n"))
  }
}
```


# PCA plots

## Principal Component Analysis
```{r}
PCA_plot(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  PCAx = params$pc_x,
  PCAy = params$pc_y,
  selected_color = params$color,
  selected_shape = params$shape,
  plots_color_select = params$plots_color_select
)
pc_factor_correlation(
  data = params$pre_processed_data,
  sample_info = params$sample_info
)
```

## Principal Component Analysis 3D
```{r fig.show='hide'}
PCA_plot_3d(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  PCAx = params$pc_x,
  PCAy = params$pc_y,
  PCAz = params$pc_z,
  selected_color = params$color,
  selected_shape = params$shape,
  plots_color_select = params$plots_color_select
)
```
Plot not included on this report.

## T-SNE
```{r}
t_SNE_plot(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  selected_color = params$color,
  selected_shape = params$shape,
  plots_color_select = params$plots_color_select
)
```

## Multi-Dimensional Scaling
```{r}
MDS_plot(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  selected_color = params$color,
  selected_shape = params$shape,
  plots_color_select = params$plots_color_select
)
```


# PCA tools tab

## Biplot
```{r}
PCA_biplot(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  select_gene_id = params$select_gene_id,
  all_gene_names = params$all_gene_names,
  selected_x = params$selected_x,
  selected_y = params$selected_y,
  encircle = params$encircle,
  showLoadings = params$showLoadings,
  pointlabs = params$pointlabs,
  point_size = params$point_size,
  ui_color = params$ui_color,
  ui_shape = params$ui_shape
)
```

## Scree Plot
```{r}
PCA_Scree(processed_data = params$pre_processed_data)
```

## Variable Importance for X-axis PC
```{r}
var_imp_plots(
  data = params$pre_processed_data,
  all_gene_names = params$all_gene_names,
  PC = params$selected_x
)
```

## Variable Importance for Y-axis PC
```{r}
var_imp_plots(
  data = params$pre_processed_data,
  all_gene_names = params$all_gene_names,
  PC = params$selected_y
)
```

## Eigenvalue Correlation
```{r}
PCAtools_eigencorplot(
  processed_data = params$pre_processed_data,
  sample_info = params$sample_info
)
```

