---
title: "iDEP: Principal Component Analysis"
output:
  html_document:
    code_folding: hide
params:
  pre_processed_data:
  pre_processed_descr:
  mapping_statistics:
  sample_info:
  pc_x:
  pc_y:
  pc_z:
  color:
  shape:
  all_gene_names:
  select_gene_id:
  selected_x:
  selected_y:
  encircle:
  showLoadings:
  pointlabs:
  point_size:
  ui_color:
  ui_shape:
  plots_color_select:
  printcode:
    label: "Display Code"
    value: TRUE
    input: checkbox
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = params$printcode,
  warning = FALSE,
  message = FALSE
)
```

<!-- This document cannot be knit directly from R studio -->

<div>

::: {style="color: Blue"}
## If this site has contributed to your work, please cite our article:

Ge, S.X., Son, E.W. & Yao, R. iDEP: an integrated web application for differential expression and pathway analysis of RNA-Seq data. BMC Bioinformatics 19, 534 (2018). <https://doi.org/10.1186/s12859-018-2486-6>
:::

##

</div>

------------------------------------------------------------------------


<!-- `r params$pre_processed_descr` -->

<!--ALL SELECTED PARAMETERS-->

```{r, include=FALSE}
for (i in 1:length(params)) {
  # exclude loaded data & sample info
  if (names(params)[i] != "sample_info" &&
      names(params)[i] != "pre_processed_data" &&
      names(params)[i] != "all_gene_names" &&
      names(params)[i] != "mapping_statistics") {
    cat(paste0(names(params)[i], ": ", params[[i]], "\n"))
  }
}
```

```{r, include=FALSE}
library(idepGolem)
```

## Summary

`r params$pre_processed_descr`

`r if (!is.null(params$mapping_statistics)) params$mapping_statistics else ""`

The 2D PCA plot displays PC`r params$pc_x` on the x-axis and PC`r params$pc_y` on the y-axis`r if (!is.null(params$color) && params$color != "Names") paste0(", colored by ", params$color) else ""`
`r if (!is.null(params$shape) && params$shape != "Names") paste0(", with point shapes representing ", params$shape) else ""`. For the biplot analysis, PC`r gsub("PC", "", params$selected_x)` and PC`r gsub("PC", "", params$selected_y)` were selected, `r if (params$encircle) "with group encircling enabled" else "without group encircling"`, `r if (params$showLoadings) "showing gene loadings" else "without gene loadings"`, and `r if (params$pointlabs) "with sample labels displayed" else "without sample labels"`. Additional dimensionality reduction methods including t-SNE and MDS were also applied for comparative visualization.


## 2D PCA Plot

```{r}
PCA_plot(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  PCAx = params$pc_x,
  PCAy = params$pc_y,
  selected_color = params$color,
  selected_shape = params$shape,
  plots_color_select = params$plots_color_select
)
```

<br>

```{r pc_correlation, results='asis'}
correlation_text <- pc_factor_correlation(
  data = params$pre_processed_data,
  sample_info = params$sample_info
)
if (!is.null(correlation_text) && nzchar(correlation_text)) {
  cat(correlation_text, "\n")
}
```

<br>

## 3D PCA Plot

```{r fig.show='hide'}
PCA_plot_3d(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  PCAx = params$pc_x,
  PCAy = params$pc_y,
  PCAz = params$pc_z,
  selected_color = params$color,
  selected_shape = params$shape,
  plots_color_select = params$plots_color_select
)
```

*Note: 3D interactive plot not displayed in static report.*

<br>

## t-SNE

```{r}
t_SNE_plot(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  selected_color = params$color,
  selected_shape = params$shape,
  plots_color_select = params$plots_color_select
)
```

<br>

## Multi-Dimensional Scaling (MDS)

```{r}
MDS_plot(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  selected_color = params$color,
  selected_shape = params$shape,
  plots_color_select = params$plots_color_select
)
```

<br>

## Biplot

```{r}
PCA_biplot(
  data = params$pre_processed_data,
  sample_info = params$sample_info,
  select_gene_id = params$select_gene_id,
  all_gene_names = params$all_gene_names,
  selected_x = params$selected_x,
  selected_y = params$selected_y,
  encircle = params$encircle,
  showLoadings = params$showLoadings,
  pointlabs = params$pointlabs,
  point_size = params$point_size,
  ui_color = params$ui_color,
  ui_shape = params$ui_shape
)
```

<br>

## Scree Plot

```{r}
PCA_Scree(processed_data = params$pre_processed_data)
```

<br>

## Variable Importance for `r params$selected_x`

```{r}
var_imp_plots(
  data = params$pre_processed_data,
  all_gene_names = params$all_gene_names,
  PC = params$selected_x
)
```

<br>

## Variable Importance for `r params$selected_y`

```{r}
var_imp_plots(
  data = params$pre_processed_data,
  all_gene_names = params$all_gene_names,
  PC = params$selected_y
)
```

<br>

## Eigenvalue Correlation

```{r}
PCAtools_eigencorplot(
  processed_data = params$pre_processed_data,
  sample_info = params$sample_info
)
```

<br>

Generated using iDEP v`r packageVersion("idepGolem")` (https://bioinformatics.sdstate.edu/idep). `r format(Sys.time(), "%Y-%m-%d %H:%M:%S %Z")`. Accuracy not guaranteed.

