---
title: "Pre-Processing and Exploratory Data Analysis"
output: 
  html_document:
    code_folding: hide
params:
  loaded_data:
  individual_data:
  descr:
  sample_info:
  all_gene_info:
  data_file_format:
  no_id_conversion:
  min_counts:
  n_min_samples_count:
  counts_transform:
  counts_log_start:
  log_transform_fpkm:
  log_start_fpkm:
  low_filter_fpkm:
  n_min_samples_fpkm:
  missing_value:
  scatter_x:
  scatter_y:
  sd_color:
  rank:
  no_fdr:
  selected_gene:
  gene_plot_box:
  use_sd:
  lab_rotate:
  plots_color_select:
  date:
    label: "Date: "
    value: !r Sys.Date()
  printcode:
    label: "Display Code"
    value: TRUE
    input: checkbox
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = params$printcode)
```

<!-- This document cannot be knit directly from R studio -->

<div>

::: {style="color: Blue"}
## If this site has contributed to your work, please cite our article:

Ge, S.X., Son, E.W. & Yao, R. iDEP: an integrated web application for differential expression and pathway analysis of RNA-Seq data. BMC Bioinformatics 19, 534 (2018). <https://doi.org/10.1186/s12859-018-2486-6>
:::

## 

</div>

------------------------------------------------------------------------


This html document contains what parameters values were selected on the IDEP interface. It also includes the plots generated from those selections.

<!-- `r params$descr` -->

<!--ALL SELECTED PARAMETERS-->

```{r, include=FALSE}
for (i in 1:length(params)) {
  # exclude loaded data & sample info
  if (names(params)[i] != "sample_info" && names(params)[i] != "loaded_data") {
    cat(paste0(names(params)[i], ": ", params[[i]], "\n"))
  }
}
```

<!-- counts_transform guide: -->

<!-- "VST: variance stabilizing transform" = 2 -->

<!-- "rlog: regularized log (slow) " = 3 -->

<!-- "EdgeR: log2(CPM+c)" = 1 -->

```{r, include = FALSE}
# devtools::load_all()
library(idepGolem)
```

## Summary
`r switch(params$data_file_format, "1" = "Read Counts data", "2" = "Normalized expression values", "3" = "Log Fold Change and corrected P values data")` were analyzed using iDEP v1.0 [[Citation](#citation-1)]. `r switch(params$data_file_format,  "1" = paste0("The data was first filtered to remove reads below ", params$min_counts," CPM in at least ", params$n_min_samples_count, " sample(s). Then the data was transformed with ",  switch(params$counts_transform, "1" = paste0("EdgeR using a pseudocount of ", params$counts_log_start), "2" = "VST: Variance Stabilizing Transformation",  "3" = "Regularized log"), ". Missing values were imputed using ", params$missing_value, "."), "2" = paste0("The data was ",switch(toString(params$log_transform_fpkm), "FALSE" = "not log transformed.", "TRUE" = paste0(" log transformed with a psuedocount of ", params$log_start_fpkm, ".")), " Then it was filtered to only keep genes above level ", params$low_filter_fpkm, " in at least ", params$n_min_samples_fpkm, " sample(s). Missing values were imputed using ", params$missing_value, "."))`

## Plots

```{r Pre-Process Data, include = FALSE }
processed_data <- idepGolem::pre_process(
  data = params$loaded_data,
  missing_value = params$missing_value,
  data_file_format = params$data_file_format,
  low_filter_fpkm = params$low_filter_fpkm,
  n_min_samples_fpkm = params$n_min_samples_fpkm,
  log_transform_fpkm = params$log_transform_fpkm,
  log_start_fpkm = params$log_start_fpkm,
  min_counts = params$min_counts,
  n_min_samples_count = params$n_min_samples_count,
  counts_transform = params$counts_transform,
  counts_log_start = params$counts_log_start,
  no_fdr = params$no_fdr
)
```

```{r }
if (params$data_file_format == 1) {
  idepGolem::total_counts_ggplot(
    counts_data = processed_data$raw_counts,
    sample_info = params$sample_info,
    type = "RAW",
    plots_color_select = params$plots_color_select
  )
}
```

```{r fig.height=5.5, fig.width=7.5 }
idepGolem::eda_scatter(
  processed_data = processed_data$data,
  plot_xaxis = params$scatter_x,
  plot_yaxis = params$scatter_y
)
```

```{r fig.height=5.5, fig.width=7.5}
idepGolem::eda_boxplot(
  processed_data = processed_data$data,
  sample_info = params$sample_info,
  plots_color_select = params$plots_color_select
)
```

```{r }
idepGolem::eda_density(
  processed_data = processed_data$data,
  sample_info = params$sample_info,
  plots_color_select = params$plots_color_select
)
```

```{r}
idepGolem::mean_sd_plot(
  processed_data = processed_data$data,
  heat_cols = params$sd_color,
  rank = params$rank
)
```

```{r}
idepGolem::gene_counts_ggplot(
  counts_data = params$loaded_data,
  sample_info = params$sample_info,
  type = "Raw",
  all_gene_info = params$all_gene_info,
  plots_color_select = params$plots_color_select
)
```

<!--  ```{r} -->
<!--   individual_plots( -->
<!--           individual_data = params$individual_data, -->
<!--           sample_info = params$sample_info,             # This is NULL, but think that's ok -->
<!--           selected_gene = params$selected_gene, -->
<!--           gene_plot_box = params$gene_plot_box, -->
<!--           use_sd = params$use_sd, -->
<!--           lab_rotate = params$angle_ind_axis_lab -->
<!--         ) -->
<!-- # ``` -->